Src/Plugins/Security/SEC101_054.AlibabaAccessKeyValidator.cs

```diff
@@ -0,0 +1,120 @@
+                                                      ref string failureLevel,
+                                                      ref string fingerprintText,
+                                                      ref string message)
+        {
+            if (!groups.TryGetNonEmptyValue("account", out string account) ||
+                !groups.TryGetNonEmptyValue("password", out string password))
+            {
+                return nameof(ValidationState.NoMatch);
+            }
+
+            fingerprintText = new Fingerprint()
+            {
+                Platform = nameof(AssetPlatform.AlibabaCloud),
+                Account = account,
+                Password = password,
+            }.ToString();
+
+            return nameof(ValidationState.Unknown);
+        }
+
+        protected override string IsValidDynamicHelper(ref string fingerprintText,
+                                                       ref string message)
+        {
+            var fingerprint = new Fingerprint(fingerprintText);
+            string account = fingerprint.Account;
+            string password = fingerprint.Password;
+
+            try
+            {
+                // Taken from https://www.alibabacloud.com/help/doc-detail/63638.htm
+                IClientProfile clientProfile = DefaultProfile.GetProfile("cn-shanghai", account, password);
+                DefaultAcsClient client = new DefaultAcsClient(clientProfile);
+
+                PubRequest request = new PubRequest();
+
+                // We don't actually care about any of the product stuff, so just leave dummy values in
+                request.ProductKey = "<productKey>";
+                request.TopicFullName = "/<productKey>/<deviceName>/get";
+                byte[] payload = Encoding.Default.GetBytes("Hello World.");
+                string payloadStr = Convert.ToBase64String(payload);
+                request.MessageContent = payloadStr;
+                request.Qos = 0;
+                PubResponse response = client.GetAcsResponse(request);
+
+                // If all goes well, we'll receive a "product format invalid" message in the response
+                // which means authentication succeeded. Therefore the id and secret are valid.
+                return ReturnAuthorizedAccess(ref message, asset: account);
+            }
+            catch (ClientException ce)
+            {
+                switch (ce.ErrorCode)
+                {
+                    case _keyNotFound:
+                        // Not even the client id we found is valid. Return no match.
+                        return nameof(ValidationState.NoMatch);
+                    case _invalidSecret:
+                        // The client ID is valid but the secret was not. We have higher confidence than
+                        // normal warnings so return authorized warning
+                        return nameof(ValidationState.AuthorizedWarning);
```

Comment:
>nameof(ValidationState.AuthorizedWarning); [](start = 31, length = 42)

I think we have a helper for this case (returnUnauthorized...., it will asset message and probably asset)

Reply:
Done

---
In reply to: [591850730](https://github.com/microsoft/sarif-pattern-matcher/pull/285#discussion_r591850730) [](ancestors = 591850730)